### PR TITLE
[many ports] Apply host dependencies

### LINKED ISSUE #16168: Error: Building package gmp:x64-windows failed with: Build_failed
Package: gmp:x64-windows
 Vcpkg version: 2021-01-13-768d8f95c9e752603d2c5901c7a7c7fbdb08af35

CMake Error at installed/x64-windows/share/yasm-tool-helper/yasm-tool-helper.cmake:16 (message):
  Cross-targetting and x64 ports requiring yasm require the x86-windows
  yasm-tool to be available.  Please install yasm-tool:x86-windows first.
Call Stack (most recent call first):
  ports/gmp/portfile.cmake:19 (yasm_tool_helper)
  scripts/ports.cmake:131 (include)

### LINKED ISSUE #16811: [protobuf] build failure
Starting package 1/5: protobuf:x86-uwp
Building package protobuf[core]:x86-uwp...
-- Using community triplet x86-uwp. This triplet configuration is not guaranteed to succeed.
-- [COMMUNITY] Loading triplet configuration from: C:\Users\haopeng\Documents\UserName\code\opensource\vcpkg\triplets\community\x86-uwp.cmake
-- Using cached C:/Users/haopeng/Documents/UserName/code/opensource/vcpkg/downloads/protocolbuffers-protobuf-2514f0bd7da7e2af1bed4c5d1b84f031c4d12c10.tar.gz
-- Cleaning sources at C:/Users/haopeng/Documents/haopeng/code/opensource/vcpkg/buildtrees/protobuf/src/31c4d12c10-2e3bb0fe91.clean. Use --editable to skip cleaning for the packages you specify.
-- Extracting source C:/Users/haopeng/Documents/haopeng/code/opensource/vcpkg/downloads/protocolbuffers-protobuf-2514f0bd7da7e2af1bed4c5d1b84f031c4d12c10.tar.gz
-- Applying patch fix-static-build.patch
-- Applying patch fix-default-proto-file-path.patch
-- Using source at C:/Users/UserName/Documents/haopeng/code/opensource/vcpkg/buildtrees/protobuf/src/31c4d12c10-2e3bb0fe91.clean
CMake Error at ports/protobuf/portfile.cmake:21 (message):
  Cross-targetting protobuf requires the x86-windows protoc to be available.
  Please install protobuf:x86-windows first.
Call Stack (most recent call first):
  scripts/ports.cmake:142 (include)

Error: Building package protobuf:x86-uwp failed with: BUILD_FAILED
Please ensure you're using the latest portfiles with `.\vcpkg update`, then
submit an issue at https://github.com/Microsoft/vcpkg/issues including:
  Package: protobuf:x86-uwp
  Vcpkg version: 2021-01-13-d67989bce1043b98092ac45996a8230a059a2d7e

Additionally, attach any relevant sections from the log files above.

@JackBoosY

### PR DESCRIPTION
**Pull request by ras0219**
Reopening of #15424.

This PR solves a long-standing issue with manifests being unable to correctly install boost while cross-compiling. This is done by enabling consumers to mark a dependency as coming from the host (via `"host": true`). The dependency can then be accessed in the portfile via `${CURRENT_HOST_INSTALLED_DIR}/subpath`.

This PR does not automatically add any subpaths of the host installed dir to helpers such as `vcpkg_configure_cmake()`. For now, consumers are intended to use `vcpkg_add_to_path()` to pick individual required executable folders:

```cmake
vcpkg_add_to_path(PREPEND "${CURRENT_HOST_INSTALLED_DIR}/tools/grpc")
```

See [docs/users/tools.md](/ras0219/vcpkg/tree/dev/roschuma/tool-ports-2/docs/users/tools.md) for more details.

### CODE REVIEW

File: docs/users/host-dependencies.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+## Producing
+
+Producing a tool has no special requirements; tools should be authored as a standard port, following all the normal policies and practices. Notably, they should build against `TARGET_TRIPLET`, not `HOST_TRIPLET` within the context of their portfile.
+
+Sometimes, it can be useful to determine whether the current context is a cross-compiling one or not. This should be done by comparing the strings `TARGET_TRIPLET` and `HOST_TRIPLET`. For example:
+
+```cmake
+string(COMPARE EQUAL "${TARGET_TRIPLET}" "${HOST_TRIPLET}" I_AM_NOT_CROSSCOMPILING)
+
```
Comment (strega-nil) [inline docs/users/host-dependencies.md:53]:
```suggestion
```
no reason to have this, imo

File: ports/yasm-tool-helper/portfile.cmake [hunk 2]
```diff
@@ -1,2 +1,2 @@
 set(VCPKG_POLICY_EMPTY_PACKAGE enabled)
-configure_file(${CMAKE_CURRENT_LIST_DIR}/yasm-tool-helper.cmake ${CURRENT_PACKAGES_DIR}/share/yasm-tool-helper/yasm-tool-helper.cmake COPYONLY)
+configure_file(${CMAKE_CURRENT_LIST_DIR}/yasm-tool-helper.cmake ${CURRENT_PACKAGES_DIR}/share/yasm-tool-helper/yasm-tool-helper.cmake @ONLY)
```
Comment (strega-nil) [inline ports/yasm-tool-helper/portfile.cmake:2]: I think that this should switch to using the solution suggested in #16478

**Review by strega-nil — COMMENTED** [review]
just some minor notes, not fully reviewed

**Comment by strega-nil** [comment]
Is there a reason that this is a draft?

**Comment by ras0219** [comment]
I wanted to ensure it builds cleanly before requesting review

**Comment by Neumann-A** [comment]
all ports using `vcpkg_configure_qmake` need an additional host dependency on `qt5-base`

**Comment by ras0219** [comment]
With this PR I'm specifically targeting this list: https://github.com/microsoft/vcpkg/blob/a44feca8c4167c9e2b7ff1d877053bc2a9ea8e9b/scripts/azure-pipelines/test-modified-ports.ps1#L139-L145

File: ports/boost-modular-build-helper/vcpkg.json [hunk 3]
```diff
@@ -1,8 +1,9 @@
 {
   "name": "boost-modular-build-helper",
   "version-string": "1.75.0",
-  "port-version": 6,
+  "port-version": 3,
```
Comment (autoantwort) [inline ports/boost-modular-build-helper/vcpkg.json:4]: Port version changes from `6` to `3`?

File: ports/boost-concept-check/vcpkg.json [hunk 4]
```diff
@@ -1,7 +1,6 @@
 {
   "name": "boost-concept-check",
   "version-string": "1.75.0",
-  "port-version": 1,
```
Comment (autoantwort) [inline ports/boost-concept-check/vcpkg.json:4]: Error?

**Review by autoantwort — COMMENTED** [review]
For the port `boost-context` and `grpc` the port version field was not updated

File: docs/users/host-dependencies.md (hunk 1 above)
Comment (autoantwort) [reply docs/users/host-dependencies.md:53]: Not useful for users, but imho useful for maintainers. If you are writing a port that depends on itself with the host flag, you have to make distinctions if you are cross compiling or not. But an example like
```cmake
if(NOT "${TARGET_TRIPLET}" STREQUAL "${HOST_TRIPLET}")
  # specific cross compile code
endif()
```
would be more helpful than the string example.

Comment (autoantwort) [reply docs/users/host-dependencies.md:53]: nvm this is the next line.

Comment (autoantwort) [reply docs/users/host-dependencies.md:53]: Is it a bug? But for me there is no variable `${HOST_TRIPLET}` or the value is the empty string?

Comment (autoantwort) [reply docs/users/host-dependencies.md:53]: > Is it a bug? But for me there is no variable ${HOST_TRIPLET} or the value is the empty string?

nwm. `set(HOST_TRIPLET "${_HOST_TRIPLET}")` gets introduced with this PR. I thought it is already there...

**Comment by Neumann-A** [comment]
another todo: all `ADD_BIN_TO_PATH` options `vcpkg_(install|build)_*` need to be disabled if `target!=host`
@ras0219-msft: I hope you keep a list with those ToDo's :)

File: docs/users/host-dependencies.md (hunk 1 above)
Comment (Neumann-A) [reply docs/users/host-dependencies.md:53]: better define `VCPKG_(IS_)CROSSCOMPILING`

**Comment by Neumann-A** [comment]
Hmm could it be that `CURRENT_HOST_INSTALLED_DIR` is not yet in vcpkg master?

**Comment by autoantwort** [comment]
> Hmm could it be that `CURRENT_HOST_INSTALLED_DIR` is not yet in vcpkg master?

Yes it gets introduced with this PR. See `scripts/ports.cmake`

**Comment by Neumann-A** [comment]
hmm probably would make sense to split this PR up in two.
a) only provide the new/extra definitions
b) fix all the ports

a) could probably be merged way faster than b)

**Comment by ras0219** [comment]
The problem is really the changes to core helpers that provoke full rebuilds if any other core changes were merged; the handful of changed ports are insignificant here

**Comment by soumyamahunt** [comment]
This PR assumes there will be only one host triplet needed, what if multiple tools need be installed with multiple host triplet? For example, what if user wants to build `contoso-code-generator` for `x86-windows` triplet and `contoso-build-system` for `x64-windows`? Although I can't think of any port that requires such configuration, I think this could cause issues in future and would be nice to see this scenario to be addressed by this PR. My proposed solution would be to allow overriding host triplet per dependency, something like this:
```json
{
    "name": "contoso-http-library",
    "version-string": "1.0.0",
    "description": "Contoso's http runtime library",
    "dependencies": [
        "contoso-core-library",
        {
            "name": "contoso-code-generator",
            "host": true,
            "host-triplet": "x86-windows"
        },
        {
            "name": "contoso-build-system",
            "host": true,
            "host-triplet": "x64-windows"
        }
    ]
}
```

Basically` host-triplet` combined with `platform` field can be used to handle any scenario for building dependencies with manifest.

**Comment by Neumann-A** [comment]
@soumyamahunt that was already discussed in https://github.com/microsoft/vcpkg/pull/15424

**Comment by JackBoosY** [comment]
We also need to modify `libvpx` dependency `yasm-tool`.

**Comment by ras0219-msft** [comment]
>another todo: all ADD_BIN_TO_PATH options vcpkg_(install|build)_* need to be disabled if target!=host

While using that is suspect, it would certainly be a breaking change to do that now. We may wish to add a warning in the future.

**Review by BillyONeal — APPROVED** [review]
